source/extensions/common/tap/tap_config_base.cc

```diff
@@ -11,14 +13,39 @@ namespace Extensions {
 namespace Common {
 namespace Tap {
 
+void Utility::addBufferToProtoBytes(envoy::data::tap::v2alpha::Body& output_bytes,
+                                    uint32_t max_buffered_bytes, const Buffer::Instance& data) {
+  const uint64_t num_slices = data.getRawSlices(nullptr, 0);
```

Comment:
Oh nice. Thanks. Forgot we had this. This will save me a bunch of time in my next PR.

Reply:
I was just looking at this, and I think if I do copyOut I'm going to need a third copy. I can move an `std::string` into the protobuf string, but I can't reserve() an std::string, copy data into it, and then set its size. Right? Is there any way to do this without another copy that I'm missing?

Reply:
You can do `resize()` the string and `copyOut` into `&str[0]`.

Reply:
some example: https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/dubbo_proxy/hessian_utils.cc#L81

Perhaps we can also make copyOut take a std::string& ?

Reply:
This is still slower since `resize()` will initialize the string data up to the length before we copy over it, but, it's not a big deal, and this will save me a bunch of time in my next PR.

Reply:
> This is still slower since `resize()` will initialize the string data up to the length.

Agree, so I think we can make copyOut take a `std::string&`, as we see this pattern being used in a couple other places.

Reply:
That still wouldn't help unless we reimplement copyout() manually outside of the libevent API. cc @brian-pane for the new buffer implementation. I think for now I will just use copyOut() and leave a TODO for perf cleanup later.

Reply:
Yeah that's what I meant, in other words, [`toString`](https://github.com/envoyproxy/envoy/blob/master/source/common/buffer/buffer_impl.cc#L204) with `start` and `size`.

Reply:
Yup agreed, will add a TODO and fix in a follow up.

Reply:
I just looked at this again and I realized that since I'm appending to the proto string across calls, this will require an extra copy no matter what we do to the buffer API, so I think I'm just going to leave it as is. I will think about what to do in my next change since this is going to get more complicated. I will add a TODO along these lines.